### PR TITLE
DeckPicker: Dismiss snackbar when it obscures FAB

### LINKED ISSUE #14470: [BUG]: Snackbar shows in front of the FAB secondary options
### Checked for duplicates?

- [X] This issue is not a duplicate

### What are the steps to reproduce this bug?

1. Do an action that shows a snackbar in the deck picker, such as undoing or syncing
2. Tap the FAB

### Expected behaviour

The snackbar is dismissed

### Actual behaviour

The snackbar is in front of the FAB secondary options (Get shared decks and Create deck)

image

### Debug info

```text
3e520ad2c64a0f5a6c7b65160825b4fd5c6d258f
```

### (Optional) Anything else you want to share?

_No response_

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by soCallmeAdityaKumar**
## Purpose / Description
Snackbar shows in front of the FAB secondary options

## Fixes
* Fixes #14470

## How Has This Been Tested?

https://github.com/ankidroid/Anki-Android/assets/101629190/c5f44b49-a2de-404b-ba12-d273e30d3d9a

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt [hunk 1]
```diff
@@ -64,6 +66,11 @@ class DeckPickerFloatingActionMenu(
         mLinearLayout.alpha = 0.5f
         mStudyOptionsFrame?.let { it.alpha = 0.5f }
         isFABOpen = true
+        if (deckPicker.syncSnackbar != null && deckPicker.syncSnackbar!!.isShown)deckPicker.syncSnackbar!!.dismiss()
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt:69]: avoid using `!!`

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt [hunk 2]
```diff
@@ -64,6 +66,11 @@ class DeckPickerFloatingActionMenu(
...
+        if (deckPicker.undoAllSnackbar != null && deckPicker.undoAllSnackbar!!.isShown)deckPicker.undoAllSnackbar!!.dismiss()
+        if (deckPicker.allDeckpickerSnackbar != null && deckPicker.allDeckpickerSnackbar!!.isShown)deckPicker.allDeckpickerSnackbar!!.dismiss()
+        if (createDeckSnackBar != null && createDeckSnackBar!!.isShown)createDeckSnackBar!!.dismiss()
+
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt:73]: not safe to use `!!`

Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt]: replaced with '?'

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt (hunk 1 above)
Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt]: replaced with '?'

**Comment by criticalAY** [comment]
The naming convention used seems a little absurd maybe other reviewer would be able ti help with that

**Comment by BrayanDSO** [comment]
I believe that you could use a `BaseSnackbarBuilderProvider` (or some similar name, I'm on my phone and can't check), store only a single reference like `shownSnackbar` then use it to dismiss the snackbar

**Review by criticalAY — APPROVED** [review]
Implementation looks fine to me

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -198,6 +198,7 @@ open class DeckPicker :
     // flag asking user to do a full sync which is used in upgrade path
     private var mRecommendFullSync = false
 
+    var baseSnackBar: Snackbar? = null
     override val baseSnackbarBuilder: SnackbarBuilder = {
         anchorView = findViewById<FloatingActionButton>(R.id.fab_main)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:203]: I meant that you could add `baseSnackbar = this` below this line and avoid the repetition along all of the activity

Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:204]: yeah thanks!!

Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:204]: @BrayanDSO @criticalAY  i think its ready to be merge please review it

**Review by criticalAY — APPROVED** [review]
Thankyou!

**Comment by soCallmeAdityaKumar** [comment]
@mikehardy @david-allison @BrayanDSO review required

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt [hunk 4]
```diff
@@ -64,6 +66,9 @@ class DeckPickerFloatingActionMenu(
         mLinearLayout.alpha = 0.5f
         mStudyOptionsFrame?.let { it.alpha = 0.5f }
         isFABOpen = true
+        if (deckPicker.baseSnackBar != null && deckPicker.baseSnackBar?.isShown == true)deckPicker.baseSnackBar?.dismiss()
+        if (createDeckSnackBar != null && createDeckSnackBar?.isShown == true)createDeckSnackBar?.dismiss()
+
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt:71]: * `deckPicker.baseSnackBar?.isShown == true` handles the case where `deckPicker.baseSnackBar == null`
* add a space after the `)` and the follow-on

```suggestion
        // ensure the snackbar doesn't obscure the menu
        if (deckPicker.baseSnackBar?.isShown == true) deckPicker.baseSnackBar?.dismiss()
        if (createDeckSnackBar?.isShown == true) createDeckSnackBar?.dismiss()
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt [hunk 5]
```diff
@@ -39,6 +40,9 @@ suspend fun FragmentActivity.undoAndShowSnackbar(duration: Int = Snackbar.LENGTH
         } else {
             TR.undoActionUndone(changes.operation)
         }
-        showSnackbar(message, duration)
+        showSnackbar(message, duration) {
+//            snackbar = this
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt:44]: don't leave commented out code

**Review by david-allison — CHANGES_REQUESTED** [review]
Outcome is great

* Could you add a `@NeedsTest` somewhere, including the issue number

A few nitpicks with code style

File: AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt (hunk 5 above)
Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt]: removed

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt (hunk 4 above)
Comment (soCallmeAdityaKumar) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt]: done

**Review by david-allison — APPROVED** [review]
Looks great, thanks!

**Comment by david-allison** [comment]
CI is failing

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 6]
```diff
@@ -479,7 +479,7 @@ abstract class AbstractFlashcardViewer :
         // despite that making no sense outside of Reviewer.kt
         currentCard = withCol {
             sched.card?.apply {
-                renderOutput()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:482]: This should be reverted

File: AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt [hunk 7]
```diff
@@ -27,6 +27,7 @@ import com.ichi2.libanki.undoableOp
 
 /** If there's an action pending in the review queue, undo it and show a snackbar */
 suspend fun FragmentActivity.undoAndShowSnackbar(duration: Int = Snackbar.LENGTH_SHORT) {
+//    var snackbar: Snackbar? = null
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Undo.kt:30]: Revert this

**Review by david-allison — COMMENTED** [review]
Brief review, I'll need to go through this on my laptop

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt [hunk 8]
```diff
@@ -50,6 +52,7 @@ class DeckPickerFloatingActionMenu(
 
     // Add Note Drawable Icon
     private val addNoteIcon: Int = R.drawable.ic_add_note
+    private var createDeckSnackBar: Snackbar? = null
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.kt:55]: This isn't used

**Review by david-allison — CHANGES_REQUESTED** [review]
This pattern means that any additional snackbars are not handled.

* move `baseSnackbar` into AnkiActivity. rename to `currentSnackbar`
* Inside the following, check if `Activity` is `AnkiActivity`, and if so, set `currentSnackbar`

https://github.com/ankidroid/Anki-Android/blob/56601041636f06e3e177749e73b75051fcb76568/AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt#L112-L116

* define `fun dismissSnackar()` inside `AnkiActivity` and call that from the FAB code

**Pull request closed without merging**